Refactor work package index styling (flexible box based solution)

myabc opened the pull request:
This PR has a broad scope and also includes:
- some refactoring of legacy styles (`main.css`)
- commit from #1605

https://www.openproject.org/work_packages/14811
https://www.openproject.org/work_packages/13498
https://www.openproject.org/work_packages/13597

as collateral:
https://www.openproject.org/work_packages/11935

ulferts committed:
places footer relatively and remove bottom padding

myabc committed:
Remove Bitters scaffolding

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Convert legacy main.css styles to Sass

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Split legacy main.css styles into separate files

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Give better names to all legacy style files

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Rename 'default' stylesheet directory to 'legacy'

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use single-line Sass comments for copyright header

Single-line comments are removed from CSS output.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Prefix legacy partial files with underscore

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove redundant .title-bar classes

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/refactor-work-package-index-styling

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Conflicts:
	app/assets/stylesheets/default/main.css.erb

myabc committed:
Reapply consolidation of avatar styles in Sass

See 874b08c1. Changes lost in previous merge.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
[wip] Add alternate layout for work package views

Absolutely position page "chrome" (top and main menus, footer).

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Lay out work package view using CSS flexible boxes

This resolves various issues with scrolling, creating 2-3 main
scrollable areas: the main menu, work packages list and, when open,
the details pane.

User Story # 14811

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'feature/non_absolute_footer' into feature/refactor-work-package-index-styling

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Conflicts:
	app/assets/stylesheets/layout/_base.sass

myabc committed:
Bring in legacy styles with require directive

Use Sprockets `require` directive rather Sass `@import directives`:
there is currently no sharing of scope (variables, mixins, etc.)
between legacy and new stylesheets. From cursory glance, this approach
appears to be slightly more performant (NOT benchmarked).

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Do not allow work packages table cells to wrap

Limit table cell width to 300px.

Bug # 13498

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Always show work package table pagination

Bug # 11935

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Always show work packages detail pane toolbar

* Rename internal details pane classes for consistency.

Bug # 13597

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Move details pane styles to correct Sass file

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Revert "[wip] Add simple, CSS3-based split view"

This solution has been superseded by solution based on CSS flexible
boxes (see 58219b1e).

This (partially) reverts commit 11d08be5c6d4d7b2695cffa3cb3c40755a6e476b.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use alt CSS border-box model in work packages view

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Quick padding fix in details pane watcher tab

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
By no means perfect, but nevertheless ready for a preliminary review.

myabc committed:
Merge branch 'dev' into feature/refactor-work-package-index-styling-flexboxes

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Conflicts:
	app/assets/stylesheets/layout/_split_view.sass

ulferts commented:
When reducing the window size by reducing the browser vertically, the menu becomes scrollable. It is desired for it to scroll vertically, but a scroll bar is also present horizontally.

ulferts commented on app/assets/stylesheets/layout/_split_view.sass:
```diff
@@ -1,313 +0,0 @@
-    width: 480px
-@media only screen and (min-width: 1440px)
-  .split-view .work-packages--details, .work-packages--split-view .work-packages--details
-    width: 580px
```
@myabc, you (accidentally?) removed the media query specifying different width for different resolutions. Now the detail pane is rather narrow on my screen.

myabc replied:
@ulferts those media query rules were moved to `work_packages.sass` (line 76).

myabc commented:
I am not able to reproduce the cuke failure locally (`cucumber features/work_packages/index_pagination.feature`)

ulferts commented on app/assets/stylesheets/content/_work_packages.sass:
```diff
@@ -27,24 +27,102 @@
 
 .work-packages--list
-  @extend %panel
+  +flex-basis(100%)
```
@myabc this causes the problems with the detail pane being too narrow. If you change this to 

<pre>
flex-grow: 1;
</pre>

it works for me (Chrome). Works when the detail pane is displayed and also when it isn't. Media query is also working. Plus I believe that this is what you want to express. 
See: http://css-tricks.com/snippets/css/a-guide-to-flexbox/

myabc replied:
@ulferts thanks. I realised last night that `flex-basis: 100%` was causing the details pane width to be ignored. I'll give this a try.

myabc committed:
Correct flexible box sizing calculation

/ HT @ulferts

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Extract properties for vertical scrolling to mixin

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Show only vertical scrollbars in main menu

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Tell Travis to test against Firefox 24

This is the minimum version we support:
https://www.openproject.org/projects/openproject/wiki/supported_browsers

See http://docs.travis-ci.com/user/firefox/

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/refactor-work-package-index-styling-flexboxes

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Conflicts:
	public/templates/work_packages.list.html

myabc committed:
Remove @import of Compass CSS3 mixins

Added erroneously in 10fc285f.

Mixin naming collisions meant vendor-prefixed flex box properties were
being generated inconsistently, breaking our layout in IE 10.

With this commit, we now rely on only one CSS mixin library: Bourbon.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Revert "Tell Travis to test against Firefox 24"

Custom versions of Firefox are not available when `sudo: false` is
specified. See d5969048.

This reverts commit 8ab66fcbc5c0c57e0a5871dfe18adc1d2cf02e15.

myabc committed:
Extract alternative/absolute layout as placeholder

Sass placeholder selectors allow us to easily apply this layout mode
with other page/per-Controller classes.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Reference absolute layout selector w/tidier syntax

This does not change the compiled output, but keeps properties grouped
in a way that (arguably) makes for better readability.

See: http://sass-lang.com/documentation/file.SASS_REFERENCE.html#parent-selector

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix width of work package view o/s project context

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Move work package layout styles to correct file

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Specify min-height for work packages split view

myabc commented:
Tested in latest Chrome, Firefox, Safari, [Firefox 24.0](https://saucelabs.com/bugs/daa2a93511b84d288daf127d28d458c0), IE 10, 11.

Also tested (and broken) in [Firefox 17](https://saucelabs.com/bugs/30e7590271e64cde9a5f3f2ca13fd017).

ulferts commented:
@myabc this looks really good :+1: 

The problem I spotted has to do with the flash messages, as they will push down everything. As long as the flash message is there, the bottom of the page (pagination & edit-button) will be pushed into the void.

ulferts commented:
Can't reproduce the failing cuke either.

ulferts commented:
@myabc another thing I noticed is that we probably need a min height for all of this. When I reduce the height very much, the elements begin to be placed over one another.

myabc committed:
[wip] Adjust height when flash message shown

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Try to fix index pagination Cuke (fails on Travis)

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@hschink would you be able to do me a favour and see if you can reproduce
the  cuke failure locally?

On Tuesday, August 19, 2014, ulferts notifications@github.com wrote:

> Can't reproduce the failing cuke either.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/opf/openproject/pull/1768#issuecomment-52658954.

manwithtwowatches commented:
@myabc Can't scroll down on the create work package view anymore

ulferts commented:
@myabc opening the context menu for a row at the bottom of the page, will open the context menu as desired. But as the context menu is opened downwards it will reach over the end of the page leading to scroll bars for the whole page. The cheap solution would be to add more space to the bottom to handle those cases. But this will destroy the layout. So I guess the menu will have to open upwards. 

![image](https://cloud.githubusercontent.com/assets/617519/3980311/71e266ae-2860-11e4-871a-dddece19c735.png)

ulferts commented:
@myabc there is less space between the tab selection and the heading on the global work packages page compared to the space on the project specific work packages page. 
![image](https://cloud.githubusercontent.com/assets/617519/3980292/422c6d10-2860-11e4-8e61-6ab6c9732157.png)

myabc committed:
Remove overly general (legacy) list styles

These styles make tabs formatting inconsistent in the outside-project-
context work package details view.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Revert "Try to fix index pagination Cuke (fails on Travis)"

This reverts commit 9ee976fc979b7ac064ecfa71a6586e79abb27024.

myabc commented:
@ulferts 

> there is less space between the tab selection and the heading on the global work packages page compared to the space on the project specific work packages page. 

should be addressed in 51bbdd5

myabc commented:
@ulferts 

> So I guess the menu will have to open upwards.

I think that solution would make the most sense, but it would be a fix that would take more than a few minutes. I'd rather deal with it as a separate WP/PR.

myabc committed:
Reduce, normalise padding o/s project context

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Mark index pagination Cuke wip (fails on Travis)

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Limit absolute layout to work packages index view

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Simplify sizing of absolute layout mode

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ulferts commented:
@myabc 

> I think that solution would make the most sense, but it would be a fix that would take more than a few minutes. I'd rather deal with it as a separate WP/PR.

Fine with that. 

I spotted another tiny mistake (sorry for getting on your nerves with those) and that also has to do with a menu:

![image](https://cloud.githubusercontent.com/assets/617519/3983763/41e1c7e8-2882-11e4-964f-405a91ae9803.png)

When the last column in the work packages table has less width than the context menu, opening the context menu will increase the width of the table (or it's container). The solution could be to place the menu differently in that case. 

Whether you want to address this now or in a separate PR is up to you. I believe there are more pressing bugs.

myabc commented:
@ulferts 

> Whether you want to address this now or in a separate PR is up to you. I believe there are more pressing bugs.

I'd prefer to make context menu adjustments in a separate PR.

ulferts commented:
@myabc space and missing scrolling fixed. :+1:

ulferts commented:
:clap:
